diff -rup gccbuilder.0//Makefile gccbuilder/Makefile
--- gccbuilder.0//Makefile	2013-03-08 13:03:09 -0600
+++ gccbuilder/Makefile	2013-03-09 05:03:36 -0600
@@ -153,8 +153,7 @@ gcc: $(GCC_BUILD)/Makefile
 	  CXXFLAGS="-mthreads -O2 -D__USE_MINGW_ACCESS" \
 	  BOOT_CXXFLAGS="-mthreads -O2 -D__USE_MINGW_ACCESS" \
 	  CXXFLAGS_FOR_TARGET="-mthreads -O2 -D__USE_MINGW_ACCESS" \
-	  LDFLAGS="-s" BOOT_LDFLAGS="-s" LDFLAGS_FOR_TARGET="-s" \
-	  bootstrap
+	  LDFLAGS="-s" BOOT_LDFLAGS="-s" LDFLAGS_FOR_TARGET="-s" 
 	cd $(GCC_BUILD)/gcc && \
 	 rm -f cc1.exe gnat1.exe cc1plus.exe f951.exe jc1.exe jvgenmain.exe \
 	  cc1obj.exe cc1objplus.exe
@@ -164,8 +163,8 @@ $(GCC_BUILD)/Makefile:
 	rm -fR $(GCC_BUILD)
 	mkdir -p $(GCC_BUILD)
 	cd $(GCC_BUILD) && \
-	 $(GCC_SRC)/configure --build=mingw32 \
-	  --enable-languages=c,c++,ada,fortran,objc,obj-c++ \
+	 $(GCC_SRC)/configure --build=mingw32 --target=i686-mingw32 \
+	  --enable-languages=c,d \
 	  --enable-threads=win32 --enable-libgomp --enable-lto \
 	  --enable-fully-dynamic-string --enable-libstdcxx-debug \
 	  --enable-version-specific-runtime-libs --with-gnu-ld \
@@ -175,7 +174,8 @@ $(GCC_BUILD)/Makefile:
 	  --enable-cxx-flags='-fno-function-sections -fno-data-sections' \
 	  --with-pkgversion="tdm-$(SUB_VER)" \
 	  $(UNWIND_ARGS) \
-	  --with-bugurl="http://tdm-gcc.tdragon.net/bugs"
+	  --with-bugurl="http://gdcproject.org/bugzilla/" \
+	  --disable-bootstrap
 
 
 ###############
@@ -420,27 +420,26 @@ gcc: $(GCC_BUILD)/Makefile
 	-export LPATH="$(GCC_PREFIX)/lib;$(GCC_PREFIX)/x86_64-w64-mingw32/lib" && \
 	 export LIBRARY_PATH=$(GCC_PREFIX)/lib/gcc/x86_64-w64-mingw32/4.6.1 && \
 	 export CPATH="$(GCC_PREFIX)/include;$(GCC_PREFIX)/x86_64-w64-mingw32/include" && \
-	 $(MAKE) -C $(GCC_BUILD) bootstrap \
-	  BOOT_CFLAGS="-O2 -m32" BOOT_LDFLAGS="-s -m32"
+	 $(MAKE) -C $(GCC_BUILD) 
 	test -f $(GCC_BUILD)/fixincludes/fixincl.o
 	test -n $(GCC_BUILD)/fixincludes/fixincl.exe
 	rm -f $(GCC_BUILD)/fixincludes/*.o
 	$(MAKE) -C $(GCC_BUILD)/fixincludes CFLAGS="-m32" LDFLAGS="-m32"
-	$(MAKE) -C $(GCC_BUILD) bootstrap \
-	 BOOT_CFLAGS="-O2 -m32" BOOT_LDFLAGS="-s -m32"
+	$(MAKE) -C $(GCC_BUILD) 
 	$(MAKE) -C $(GCC_BUILD) install prefix=$(GCC_STAGE)
 $(GCC_BUILD)/Makefile:
 	rm -fR $(GCC_BUILD)
 	mkdir -p $(GCC_BUILD)
 	cd $(GCC_BUILD) && \
 	 $(GCC_SRC)/configure --build=x86_64-w64-mingw32 --enable-targets=all \
-	  --enable-languages=c,c++,fortran,objc,obj-c++ --enable-libgomp \
+	  --enable-languages=c,d --enable-libgomp \
 	  --enable-lto --enable-libstdcxx-debug \
 	  --enable-version-specific-runtime-libs --enable-fully-dynamic-string \
 	  --with-gnu-ld --disable-werror --disable-nls --disable-win32-registry \
 	  --prefix=$(GCC_PREFIX) --with-local-prefix=$(GCC_PREFIX) \
 	  --with-pkgversion="tdm64-$(SUB_VER)" \
-	  --with-bugurl="http://tdm-gcc.tdragon.net/bugs"
+	  --with-bugurl="http://gdcproject.org/bugzilla/"
+	  --disable-bootstrap
 
 .PHONY: expat
 expat: $(EXPAT_BUILD)/Makefile
Only in gccbuilder.0/: Makefile.orig
diff -rup gccbuilder.0//build-tdm-sjlj.sh gccbuilder/build-tdm-sjlj.sh
--- gccbuilder.0//build-tdm-sjlj.sh	2013-03-08 13:03:10 -0600
+++ gccbuilder/build-tdm-sjlj.sh	2013-03-08 13:03:17 -0600
@@ -3,7 +3,7 @@
 make -f /crossdev/gccbuilder/Makefile \
  JOBCOUNT=8 \
  EDITION=tdm32 \
- GCC_PREFIX=/mingw32tdm \
+ GCC_PREFIX=/crossdev/MinGW32/ \
  GMP_SRC=/crossdev/src/gmp-4.3.2 \
  GMP_BUILD=/crossdev/build/gmp-tdm32 \
  GMP_STAGE=/crossdev/support-stage-tdm32/gmp \
Only in gccbuilder.0/: build-tdm-sjlj.sh.orig
diff -rup gccbuilder.0//build-tdm64.sh gccbuilder/build-tdm64.sh
--- gccbuilder.0//build-tdm64.sh	2013-03-08 13:03:11 -0600
+++ gccbuilder/build-tdm64.sh	2013-03-08 13:04:44 -0600
@@ -1,9 +1,8 @@
 #!/bin/sh
 
 make -f /crossdev/gccbuilder/Makefile \
- JOBCOUNT=8 \
  EDITION=tdm64 \
- GCC_PREFIX=/mingw64tdm \
+ GCC_PREFIX=/crossdev/MinGW64/ \
  BINUTILS_SRC=/crossdev/src/binutils-2.21.53-20110731 \
  BINUTILS_BUILD=/crossdev/build/binutils-tdm64 \
  BINUTILS_STAGE=/crossdev/binutils-stage-tdm64 \
Only in gccbuilder.0/: build-tdm64.sh.orig
