diff -rup mingw64-runtime-svn4483-src.original//mingw-w64-crt/crt/tlssup.c mingw64-runtime-svn4483-src/mingw-w64-crt/crt/tlssup.c
--- mingw64-runtime-svn4483-src.original//mingw-w64-crt/crt/tlssup.c	2011-12-19 02:31:35 -0500
+++ mingw64-runtime-svn4483-src/mingw-w64-crt/crt/tlssup.c	2011-12-19 02:32:55 -0500
@@ -50,8 +50,8 @@ _CRTALLOC(".tls$ZZZ") char _tls_end = 0;
 _CRTALLOC(".CRT$XLA") PIMAGE_TLS_CALLBACK __xl_a = 0;
 _CRTALLOC(".CRT$XLZ") PIMAGE_TLS_CALLBACK __xl_z = 0;
 
-_CRTALLOC(".tls") const IMAGE_TLS_DIRECTORY _tls_used = {
-  (ULONG_PTR) &_tls_start+1, (ULONG_PTR) &_tls_end,
+_CRTALLOC(".rdata$TLS") const IMAGE_TLS_DIRECTORY _tls_used = {
+  (ULONG_PTR) &_tls_start, (ULONG_PTR) &_tls_end,
   (ULONG_PTR) &_tls_index, (ULONG_PTR) (&__xl_a+1),
   (ULONG) 0, (ULONG) 0
 };
